delta-app/src/main/java/io/cdap/delta/store/StateStoreMigrator.java

```diff
@@ -0,0 +1,116 @@
+  private final StateStore remoteStateStore;
+
+  private FileSystem fileSystem;
+  private Path instanceBasePath;
+  protected HCFSStateStore hcfsStateStore;
+  protected DeltaWorkerId previousId;
+
+  public StateStoreMigrator(String hcfsStateStorePath, DeltaWorkerId currentId, StateStore remoteStateStore) {
+    this.hcfsStateStorePath = hcfsStateStorePath;
+    this.currentId = currentId;
+    this.remoteStateStore = remoteStateStore;
+  }
+
+  public void perform() throws IOException {
+    //Check if path is not empty
+    if (hcfsStateStorePath == null || hcfsStateStorePath.isEmpty()) {
+      return;
+    }
+
+    //Initiate HCFS state store to read
+    Path path = new Path(hcfsStateStorePath);
+    hcfsStateStore = HCFSStateStore.from(path);
+
+    //Get previous latest generation : required as we cannot access previous context.
+    Long previousGeneration = hcfsStateStore.getLatestGeneration(currentId.getPipelineId().getNamespace(),
+                                                                 currentId.getPipelineId().getApp());
+    if (previousGeneration == null) {
+      return;
+    }
+
+    previousId = new DeltaWorkerId(new DeltaPipelineId(currentId.getPipelineId().getNamespace(),
+                          currentId.getPipelineId().getApp(), previousGeneration), currentId.getInstanceId());
+
+    fileSystem = path.getFileSystem(new Configuration());
+    //Base path for this worker
+    instanceBasePath = new Path(path, String.format("%s/%s/%d/%d",
+                                                    previousId.getPipelineId().getNamespace(),
+                                                    previousId.getPipelineId().getApp(),
+                                                    previousId.getPipelineId().getGeneration(),
+                                                    previousId.getInstanceId()));
+
+    if (!fileSystem.exists(instanceBasePath)) {
+      return;
+    }
+
+    FileStatus[] fStats = fileSystem.listStatus(instanceBasePath);
+    if (fStats.length == 0) {
+      return;
+    }
+
+    for (FileStatus stateFile : fStats) {
+      Path filePath = stateFile.getPath();
+
+      if (filePath.getName().equals(RemoteStateStore.OFFSET_KEY)) {
+        migrateOffset();
+      } else {
+        migrateStateFile(filePath.getName());
+      }
+    }
+
+    fileSystem.rename(instanceBasePath, new Path(path, String.format("%s/%s/%d/%s",
+                                                                     previousId.getPipelineId().getNamespace(),
+                                                                     previousId.getPipelineId().getApp(),
+                                                                     previousId.getPipelineId().getGeneration(),
+                                                                     previousId.getInstanceId() + "_migrated")));
+  }
+
+  private void migrateOffset() throws IOException {
+    if (remoteStateStore.readOffset(this.currentId) == null) {
```

Comment:
Do we need this if condition? migration will only happen once right? after that files are renamed so migrateOffset should not get called again?

Reply:
So this is to mitigate the case when 
- We have upserted data into SQL , but before the folder is renamed there is a failure. (although in next try , the same data should be re written) 
- Also in future incase the folder folder is renamed back to original name, then on restart it might overwrite the newer values in DB. 

It's just a safety check to enforce only write once.